### PR TITLE
Integrating repo_policy_compliance to run before runner start job

### PR DESCRIPTION
**Pull request by yhaliaw**
Add repo policy compliance as a HTTP service managed by the charm.
The runners during pre-run will call the service to resolve whether the job is valid or not.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -500,6 +525,52 @@ def _install_deps() -> None:
...
+
+        service_token_path = Path("token")
+        repo_check_web_service_path = Path("/home/ubuntu/repo_policy_compliance_service")
+        repo_check_web_service_script = Path("src/repo_policy_compliance_service.py")
+        repo_check_systemd_service = Path("/etc/systemd/system/repo-policy-compliance.service")
+
+        logger.info("Getting the secret token...")
+        if service_token_path.exists():
+            logger.info("Found existing token file.")
+            with open(service_token_path, "r") as file:
```
Comment (jdkandersson) [inline src/charm.py:541]: Pathlib.read_text could be helpful here, it is a one-liner to do this

File: src/charm.py [hunk 2]
```diff
@@ -500,6 +525,52 @@ def _install_deps() -> None:
...
+                self.service_token = file.read()
+        else:
+            logger.info("Generate new token.")
+            self.service_token = secrets.token_hex(16)
+            with open(service_token_path, "w") as file:
```
Comment (jdkandersson) [inline src/charm.py:546]: Pathlib.write_text does this in one line

File: src/charm.py [hunk 3]
```diff
@@ -472,13 +482,27 @@ def _reconcile_runners(self, runner_manager: RunnerManager) -> Dict[str, "JsonOb
...
             return {"delta": {"virtual-machines": 0}}
 
-    @staticmethod
     @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
-    def _install_deps() -> None:
+    def _install_deps(self) -> None:
         """Install dependencies."""
         logger.info("Installing charm dependencies.")
 
         # Binding for snap, apt, and lxd init commands are not available so subprocess.run used.
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply src/charm.py:490]: This looks good, I wasn't aware. I propose that we include the use of this library in the upcoming refactor of this charm

[the remark this answers is not in the document]
Comment (mthaddon) [reply src/charm.py:495]: That sounds good to me, I'll add a note to the relevant task.

**Review by gregory-schiano — APPROVED** [review]
LGTM

**Review by weiiwang01 — APPROVED** [review]

**Pull request merged**